### PR TITLE
Fix overload resolution on Tuples

### LINKED ISSUE #1578: Explicit List of Tuple do not support __setitem__ with index
```python
from typing import List, Tuple

l = []  # type: List[Tuple[int, int]]
l.append((1, 2))

l[0] = 3, 4
l[:] = [(5, 6)]
```

gives

```
asd.py:6: error: No overload variant of "__setitem__" of "list" matches argument types [builtins.int, Tuple[builtins.int, builtins.int]]
```
so `__setitem__` with slice works but not with index.

Not specifing the type does work (it is just an example, the code I'm working on need to explicit a specific variable).

--

I don't think it is an issue in typeshed as playing with builtins.list do not change a thing.
btw, simply exchanging the `@overload  __setitem__` gives me weirder errors, like
```
asd.py:1: note: In module imported here:
/home/tharvik/mypy/typeshed/stdlib/3/builtins.pyi: note: In class "list":
/home/tharvik/mypy/typeshed/stdlib/3/builtins.pyi:477: error: Signature of "__setitem__" incompatible with supertype "MutableSequence"
asd.py: note: At top level:
asd.py:6: error: No overload variant of "__setitem__" of "list" matches argument types [builtins.int, Tuple[builtins.int, builtins.int]]
```
The order shouldn't matter, right? (I didn't see it written in the PEP anyway)

### PR DESCRIPTION
**Pull request by rwbarton**
Fixes #1578, since list's __setitem__ is overloaded.

### CODE REVIEW

File: mypy/test/data/check-overloading.test [hunk 1]
```diff
@@ -678,3 +678,16 @@ def f(x: int) -> None: pass
...
 @overload
 def f(x: A) -> int: pass
+
+[case testOverloadTuple]
+from typing import overload, Tuple
+@overload
+def f(x: int, y: Tuple[str, ...]) -> None: pass
+@overload
+def f(x: int, y: str) -> None: pass
+f(1, ('2', '3'))
```
Comment (JukkaL) [inline mypy/test/data/check-overloading.test]: Also test calling with a `Tuple[str, ...]` arg and a `Tuple[int, ...]` argument.

Comment (rwbarton) [inline mypy/test/data/check-overloading.test]: Added.

Comment (rwbarton) [inline mypy/test/data/check-overloading.test]: Oops, now it's actually added.

**Comment by JukkaL** [comment]
@rwbarton Feel free to land this PR at any time if tests pass after fixing rebase conflicts. Sorry for the delay!

**Comment by rwbarton** [comment]
Hmm, I thought I commented somewhere about this PR being unneeded after #1596, but now I can't find that comment...

**Comment by rwbarton** [comment]
Oh right, I remember now.
```py
    # Fall back to a conservative equality check for the remaining kinds of type.
    return 2 if is_same_type(erasetype.erase_type(actual), erasetype.erase_type(formal)) else 0
```
With the change to `erase_type` in #1596, this now correctly handles the case of calling a Tuple formal argument with a Tuple value, so the workaround in this PR is no longer needed.

I'll still merge the tests though since they are a bit different than the ones added in #1596.

**Comment by rwbarton** [comment]
I added the tests in commit https://github.com/python/mypy/commit/838b4e0b69149d3214d0cc730829ee349bc5aeab.

**Pull request closed without merging**